cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go

```diff
@@ -362,32 +362,39 @@ func ValidateKubeconfigsForExternalCA(outDir string, cfg *kubeadmapi.InitConfigu
 }
 
 func getKubeConfigSpecsBase(cfg *kubeadmapi.InitConfiguration) (map[string]*kubeConfigSpec, error) {
-	apiServer, err := kubeadmutil.GetControlPlaneEndpoint(cfg.ControlPlaneEndpoint, &cfg.LocalAPIEndpoint)
+	controlPlaneEndpoint, err := kubeadmutil.GetControlPlaneEndpoint(cfg.ControlPlaneEndpoint, &cfg.LocalAPIEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	localAPIEndpointIP, localAPIEndpointPort, err := kubeadmutil.ParseAPIEndpoint(&cfg.LocalAPIEndpoint)
 	if err != nil {
 		return nil, err
 	}
+	localAPIEndpoint := kubeadmutil.FormatURL(localAPIEndpointIP.String(), localAPIEndpointPort).String()
+
 	return map[string]*kubeConfigSpec{
 		kubeadmconstants.AdminKubeConfigFileName: {
-			APIServer:  apiServer,
+			APIServer:  controlPlaneEndpoint,
 			ClientName: "kubernetes-admin",
 			ClientCertAuth: &clientCertAuth{
 				Organizations: []string{kubeadmconstants.SystemPrivilegedGroup},
 			},
 		},
 		kubeadmconstants.KubeletKubeConfigFileName: {
-			APIServer:  apiServer,
+			APIServer:  controlPlaneEndpoint,
```

Comment:
for control plane nodes, it might make sense for the kubelet to speak to the local API server (e.g. dynamic environments, machines that are using DHCP where the lease expire)

Reply:
From the issue I understood this is going to be tackled in a separated PR, is this right @neolit123

Reply:
that is true and we spoke about this yesterday.

the problem is that kubeadm join has a kubelet-start phase that blocks until it connects to an api-server at the CPE.
https://github.com/kubernetes/kubernetes/blob/f2e3154a140dc192760a5cf8c01f8044e1aa867b/cmd/kubeadm/app/cmd/phases/join/kubelet.go#L99

https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join/
the CP manifests + etcd for this node are written after the TLS bootstrap.

this implies rework of the kubeadm join phases and needs some discussion (maybe tomorrow @ office hours).

Reply:
sounds good, I'll join

Reply:
i will make this PR not close the issue and we can have multiple checkboxes in it to leave time for investigating the kubelet.conf optional change.

Reply:
done.

Reply:
+1 to both using the local address for the kubelet and also treating it as a separate PR, this will help to greatly simplify the bootstrapping process when using a load balancer, since the load balancer will not need to be pre-created for bootstrapping.